### ISSUE TITLE
Support Write on Virtual Dataset

### ISSUE THREAD

**Issue by jisonZ** [h5py/h5py#1798]
Hi! I'm currently working on a project that involves a possible write on HDF5 virtual dataset, specifically writing on the VDS will also write to the files that constructed the VDS. I tried the following but doesn't seem to work:
`f = h5py.File("test.h5", 'w')
f["vds"][4][3][2][1] = 1`
Suppose the 'test.h5' already have a 4d VDS called "vds" , this assigment did not modify the VDS itself, left alone the files behind it.
To be honest, I didn't look into the C version of HDF5 API, but from the documentation, it seems that they support this functionality (not too sure) I wonder if h5py does support this feature that I missed some info from the documentation or it's not supported at all.

**Comment by tacaswell** [comment]
The first thing we need to sort out is if the hdf5 group thinks this is supported at the library level.  We try to stick very closely to mirroring out the underlying functionality (in a pythonic way!).

**Comment by jisonZ** [comment]
@tacaswell I agree. I could take a closer look at the HDF5 library and follow up on the discussion

**Comment by takluyver** [comment]
Note that the `ds[4][3][2][1] =` won't work for writing to a normal dataset either - this section of the docs explains why:

https://docs.h5py.org/en/stable/high/dataset.html#multiple-indexing

**Comment by takluyver** [comment]
I can't find any definitive documentation on whether it's supposed to work (the RFC documentimplies it should, but that was written before it was implemented, and they acknowledge it doesn't exactly match reality). When I tried with h5py, it appears to work as expected (modifying the source datasets) if I match the dtype. Using the dataset created by the [`vds_simple.py`](https://github.com/h5py/h5py/blob/master/examples/vds_simple.py) example, I did:

```python
vd[:, :5] = np.arange(20).reshape(4, 5).astype(np.int32)
```

However, it doesn't seem to do type conversions correctly. If I leave off the `astype()` there, it writes int64 data instead of int32, and you get numbers up to 9 interspersed with 0s instead of numbers up to 19.

I'll ask on the HDF forum if this is meant to work or not.

**Comment by takluyver** [comment]
Asked about it here: https://forum.hdfgroup.org/t/writing-to-virtual-datasets/8188
